openslides/users/views.py

```diff
@@ -524,3 +528,113 @@ def post(self, request, *args, **kwargs):
         else:
             raise ValidationError({'detail': _('Old password does not match.')})
         return super().post(request, *args, **kwargs)
+
+
+class PasswordResetView(APIView):
+    """
+    Users can send an email to themselves to get a password reset email.
+
+    Send POST request with {'email': <email addresss>} and all users with this
+    address will receive an email (means Django sends one or more emails to
+    this address) with a one-use only link.
+    """
+    http_method_names = ['post']
+    use_https = False  # TODO: Do we use https?
+
+    def post(self, request, *args, **kwargs):
+        """
+        Loop over all users and send emails.
+        """
+        to_email = request.data.get('email')
+        for user in self.get_users(to_email):
+            current_site = get_current_site(request)
+            site_name = current_site.name
+            context = {
+                'email': to_email,
+                'site_name': site_name,
+                'protocol': 'https' if self.use_https else 'http',
+                'domain': current_site.domain,
+                'path': '/login/reset-password-confirm/',
+                'user_id': urlsafe_base64_encode(force_bytes(user.pk)).decode(),
+                'token': default_token_generator.make_token(user),
+                'username': user.get_username(),
+            }
+            # Send a django.core.mail.EmailMessage to `to_email`.
+            subject = _('Password reset for {}').format(site_name)
+            subject = ''.join(subject.splitlines())
+            body = self.get_email_body(**context)
+            from_email = None  # TODO: Add nice from_email here.
+            email_message = mail.EmailMessage(subject, body, from_email, [to_email])
+            email_message.send()
+        return super().post(request, *args, **kwargs)
+
+    def get_users(self, email):
+        """Given an email, return matching user(s) who should receive a reset.
+
+        This allows subclasses to more easily customize the default policies
+        that prevent inactive users and users with unusable passwords from
+        resetting their password.
+        """
+        active_users = User.objects.filter(**{
+            'email__iexact': email,
+            'is_active': True,
+        })
+        return (u for u in active_users if u.has_usable_password())
+
+    def get_email_body(self, **context):
+        """
+        Add context to email template and return the complete body.
+        """
+        return textwrap.dedent(
+            """
+            You're receiving this email because you requested a password reset for your user account at {site_name}.
+
+            Please go to the following page and choose a new password:
+
+            {protocol}://{domain}{path}?user_id={user_id}&token={token}
+
+            Your username, in case you've forgotten: {username}
```

Comment:
including the username in the E-Mail is a security flaw.
Intercepting the E-Mail would give you full access.
Knowing the E-Mail address of a registered administrator is sufficient to take over the software.

Reply:
@normanjaeckel Ping

Reply:
This is the django default text. If someone lost his username he cannot reset his account. @emanuelschuetze What do you think? (It wouldn't be a problem for me if you change the text.

Reply:
Since it's possible to enter an email twice that should be OK